### PR TITLE
Do less work in the inner loop of `assign_commits`

### PR DESCRIPTION
**Pull request by sgrif**
The release is always the same, so we don't need to do it in a loop. We
can get the author name and email from `git log` so we don't need to
subshell in a loop. Finally, rather than selecting a sha and trying to
update it if it exists, we can just try to update the sha and insert if
nothing was updated.

With Diesel 0.12, we can actually do this in 2 queries total by doing
something along the lines of the following:

```rust
let mut new_authors: Vec<NewAuthor> = commits.map(...);
new_authors.dedup();
let authors = insert(&new_authors.on_conflict((name, email), do_update().set(id.eq(id)))) // dummy update to get the row to return
    .into(authors::table)
    .get_results::<Author>(&connection)?;
let author_ids = authors.into_iter()
    .map(|author| ((author.name, author.email), author.id))
    .collect::<HashMap<_, _>>();

let commits = commits.map(...); // get author id from map above
let commits = insert(&new_commits.on_conflict(sha, do_update().set(commits::release_id.eq(the_release.id))))
    .into(commits::table)
    .execute(&connection)
```

But we'd lose the logging and I'm guessing performance isn't that
critical here anyway.

### CODE REVIEW

**Comment by steveklabnik** [comment]
> But we'd lose the logging and I'm guessing performance isn't that
critical here anyway.

I'm not super worried about the logging, and honestly, this is the part of the process that takes the most time. We have a ticket that's related even https://github.com/rust-lang-nursery/thanks/issues/34

in general stuff is _super_ not optimized here

**Comment by sgrif** [comment]
Oh well in that case I can optimize this down to 4 queries with Diesel 0.11.

**Comment by steveklabnik** [comment]
I'm also happy to upgrade diesel, looks like 0.12 isn't out yet though

**Comment by sgrif** [comment]
No, it's not. https://github.com/diesel-rs/diesel/pull/793 is the PR to add `ON CONFLICT DO UPDATE`. We can just do an `INSERT ON CONFLICT DO NOTHING` followed by a select though.

**Comment by sgrif** [comment]
Ok I've updated this to do the whole thing in 4 commits. It compiles, but maybe test it against a local dataset first or backup the prod database since I haven't tested it.

File: src/releases.rs [hunk 1]
```diff
@@ -33,71 +38,83 @@ pub fn assign_commits(log: &Logger, release_name: &str, previous_release: &str,
...
+{
+    let new_authors = git_log.iter().map(|&(_, email, name)| {
+        NewAuthor { email: email, name: name }
+    }).collect();
+    let author_ids = ::authors::find_or_create_all(conn, new_authors)?
+        .into_iter()
+        .map(|author| ((author.email, author.name), author.id))
+        .collect::<HashMap<_, _>>();
+    Ok(git_log.into_iter()
+        .map(|(sha, email, name)| (sha, author_ids[&(email.into(), name.into())]))
```
Comment (sgrif) [inline src/releases.rs]: I'm really mad that I couldn't index a `HashMap<(String, String)` with a value of `(&str, &str)`

**Comment by steveklabnik** [comment]
```
thread 'main' panicked at 'Could not find or create author: DatabaseError(__Unknown, "syntax error at or near \"ON\"")',
```

**Comment by steveklabnik** [comment]
Ah I have 9.3 locally, i should be running the same as heroku, `9.6.1`. will test later

**Comment by steveklabnik** [comment]
I finally got this all fixed:

```text
Mar 13 10:38:01.748 INFO Assigning commits to release 0.2, version: 0.1.0
thread 'main' panicked at 'Error saving commits and authors: StringError("Expected to create or update 1556 commits, but only 3112 were")', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\result.rs:837
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

so yeah, not sure whats up with that?

**Comment by sgrif** [comment]
:eyes:

**Comment by sgrif** [comment]
Ok try again with that migration

**Comment by steveklabnik** [comment]
Farther!

```text
Mar 13 11:19:44.711 INFO Assigning commits to release 1.0.0, version: 0.1.0
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\option.rs:323
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

not sure what's up with that. It works for the first number of releases:

```text
Mar 13 11:19:38.132 INFO Creating commit: c01efc669f09508b55eced32d3c88702578a7c3e, version: 0.1.0
Mar 13 11:19:38.167 INFO Assigning commits to release 0.2, version: 0.1.0
Mar 13 11:19:38.526 INFO Assigning commits to release 0.3, version: 0.1.0
Mar 13 11:19:38.985 INFO Assigning commits to release 0.4, version: 0.1.0
Mar 13 11:19:39.427 INFO Assigning commits to release 0.5, version: 0.1.0
Mar 13 11:19:39.696 INFO Assigning commits to release 0.6, version: 0.1.0
Mar 13 11:19:40.235 INFO Assigning commits to release 0.7, version: 0.1.0
Mar 13 11:19:40.773 INFO Assigning commits to release 0.8, version: 0.1.0
Mar 13 11:19:41.404 INFO Assigning commits to release 0.9, version: 0.1.0
Mar 13 11:19:41.832 INFO Assigning commits to release 0.10, version: 0.1.0
Mar 13 11:19:42.244 INFO Assigning commits to release 0.11.0, version: 0.1.0
Mar 13 11:19:42.728 INFO Assigning commits to release 0.12.0, version: 0.1.0
Mar 13 11:19:43.216 INFO Assigning commits to release 1.0.0-alpha, version: 0.1.0
Mar 13 11:19:43.904 INFO Assigning commits to release 1.0.0-alpha.2, version: 0.1.0
Mar 13 11:19:44.350 INFO Assigning commits to release 1.0.0-beta, version: 0.1.0
Mar 13 11:19:44.711 INFO Assigning commits to release 1.0.0, version: 0.1.0
```

**Comment by steveklabnik** [comment]
`\src\releases.rs:58`

**Comment by sgrif** [comment]
Ok well hopefully that commit fixed it

**Comment by steveklabnik** [comment]
```
Mar 13 11:57:17.885 INFO Assigning commits to release 1.0.0, version: 0.1.0
thread 'main' panicked at 'Tried to read column names from empty list of rows', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\option.rs:715
```

panic is here:

```
  18:     0x7ff66017ec7f - thanks::authors::find_or_create_all
                        at C:\Users\steve\src\thanks\src\authors.rs:27
  19:     0x7ff66017dfd3 - thanks::releases::authors_by_sha
                        at C:\Users\steve\src\thanks\src\releases.rs:111
  20:     0x7ff66017d3ad - thanks::releases::assign_commits::{{closure}}
                        at C:\Users\steve\src\thanks\src\releases.rs:66
```

**Comment by sgrif** [comment]
Ugh.. I guess https://github.com/diesel-rs/diesel/issues/384 is back... https://github.com/diesel-rs/diesel/issues/797

Is there a release with 0 commits or something?

**Comment by steveklabnik** [comment]
> Is there a release with 0 commits or something?

I don't think so? The smallest should be the `x.y.z` releases, which have like two or three commits.

**Comment by steveklabnik** [comment]
.. and given that this is crashing out right after saying "do 1.0.0," that one should def. not have zero commits 😄

**Comment by sgrif** [comment]
What's the command for me to try this locally?

**Comment by sgrif** [comment]
Oh I guess it's in the readme

**Comment by steveklabnik** [comment]
Yup, the populate command.

**Comment by sgrif** [comment]
For me it crashes on 1.14.0 not 1.0.0, and it looks like it's because the tag was missing locally. `git fetch --tags` fixed it.

**Comment by sgrif** [comment]
Also I'm pretty sure that "create all commits before assigning" loop that `bin/populate` does is 100% redundant with the work that `assign_releases` does

**Comment by steveklabnik** [comment]
> Also I'm pretty sure that "create all commits before assigning" loop that bin/populate does is 100% redundant with the work that assign_releases does

Yes this is a long-standing "look into that" thing for me.

Let me try again with tags fetched...

**Comment by steveklabnik** [comment]
Wow. yeah. Here's what happened: My clone of rust is fairly recent, since before last time I did any work on thanks, and git doesn't automatically fetch all tags when you clone; you need an explicit `--tags`. ugh. sorry about that.

**Comment by sgrif** [comment]
No worries. I'm updating the code to actually check that `git log` exited successfully, too

**Comment by steveklabnik** [comment]
Awesome. So this is ready to :shipit: now, yeah?

**Comment by sgrif** [comment]
Yes

**Comment by steveklabnik** [comment]
🤘 ❤️
